File: environment/test_launcher.py
import importlib.util
from pathlib import Path
import subprocess
import tempfile
import unittest
from unittest.mock import patch

import yaml

spec = importlib.util.spec_from_file_location('launcher', Path(__file__).with_name('launcher.py'))
launcher = importlib.util.module_from_spec(spec)
spec.loader.exec_module(launcher)


class RecipeTests(unittest.TestCase):
    def setUp(self):
        self.tmp = tempfile.TemporaryDirectory()
        self.addCleanup(self.tmp.cleanup)
        self.directory = Path(self.tmp.name).resolve()

    def load(self, data):
        file = self.directory / 'config.yaml'
        file.write_text(yaml.safe_dump(data))
        return launcher.load_config(file)

    def test_relative_policy_selection_and_preservation_defaults(self):
        (self.directory / 'requirements.toml').write_text('[features]\napps = false\n')
        result = self.load({'codex_requirements': 'requirements.toml'})
        self.assertEqual(result['codex_policy_mode'], 'custom')
        self.assertEqual(result['claude_policy_mode'], 'preserve')
        self.assertEqual(result['codex_requirements'], str(self.directory / 'requirements.toml'))

    def test_reject_invalid_inputs_before_vm_access(self):
        for data in [
            {'ansible_connection': 'local'}, {'packages': 'curl'},
            {'environment': {'INVALID-NAME': 'value'}},
            {'dotfiles_install': '../escape'}, {'dotfiles_repo': '--upload-pack=bad'},
            {'replace_codex_config': True}, {'reset_codex_requirements': 'false'},
            {'credentials': [{'name': 'TOKEN', 'description': 'token'}], 'environment': {'TOKEN': 'bad'}},
        ]:
            with self.subTest(data=data), self.assertRaises(ValueError):
                self.load(data)

    def test_invalid_custom_settings(self):
        (self.directory / 'settings.json').write_text('{"sandbox":{"enabled":"yes"}}')
        with self.assertRaises(Exception):
            self.load({'claude_config': 'settings.json'})
        (self.directory / 'bad.toml').write_text('[features]\napps = "yes"\n')
        with self.assertRaises(ValueError):
            self.load({'codex_requirements': 'bad.toml'})
        (self.directory / 'managed.json').write_text('{"sandbox":{"enabled":"yes"}}')
        with self.assertRaises(launcher.jsonschema.ValidationError):
            self.load({'claude_managed_settings': 'managed.json'})

    def test_ssh_install_preserves_existing_config_and_is_repeatable(self):
        state = {'name': 'test', 'dir': '/tmp/lima directory/test'}
        (self.directory / '.ssh').mkdir()
        config = self.directory / '.ssh/config'
        original = 'Host personal\n  HostName example.com\n'
        config.write_text(original)
        with patch.object(launcher.Path, 'home', return_value=self.directory):
            launcher.install_ssh(state)
            launcher.install_ssh(state)
        self.assertEqual(config.read_text().count('Include'), 1)
        self.assertTrue(config.read_text().endswith(original))
        self.assertEqual(len(list(config.parent.glob('config.before-*'))), 1)
        generated = (config.parent / 'lima-environments/test.config').read_text()
        self.assertIn('ControlPath ~/.ssh/control-%C', generated)
        self.assertLess(generated.index('IdentityAgent none'), generated.index('Include'))


class BoundaryTests(unittest.TestCase):
    def state(self):
        return {'name': 'test', 'dir': '/tmp/lima/test', 'config': {
            'plain': True, 'mounts': [], 'provision': [],
            'ssh': {'forwardAgent': False, 'loadDotSSHPubKeys': False},
            'user': {'name': 'dev', 'home': '/home/dev', 'uid': 1000}}}

    def test_global_overrides_fail_closed(self):
        for key, value in [('mounts', [{'location': '~'}]), ('provision', [{'script': 'bad'}]),
                           ('plain', False), ('propagateProxyEnv', True),
                           ('portForwards', [{'static': True}])]:
            state = self.state()
            state['config'][key] = value
            with self.subTest(key=key), self.assertRaises(ValueError):
                launcher.check_instance(state, 'test')
        launcher.check_instance(self.state(), 'test')

    def test_ssh_separates_root_and_dev_and_disables_agent(self):
        with patch.object(launcher, 'run') as run:
            for user in ['dev', 'root']:
                launcher.ssh(self.state(), user, ['echo', 'literal $(not-a-command)'])
        for call, user in zip(run.call_args_list, ['dev', 'root']):
            args = call.args[0]
            self.assertEqual(args[args.index('-l') + 1], user)
            self.assertIn('ControlPath=~/.ssh/control-%C', args)
            self.assertIn('IdentityAgent=none', args)
            self.assertEqual(args[-1], "echo 'literal $(not-a-command)'")

    def test_missing_credentials_require_terminal_without_prompting_or_writing(self):
        with patch.object(launcher, 'ssh', return_value=subprocess.CompletedProcess([], 0, '[]')) as ssh:
            with patch.object(launcher.sys.stdin, 'isatty', return_value=False):
                with self.assertRaises(ValueError):
                    launcher.credentials(self.state(), [{'name': 'TOKEN', 'description': 'test'}])
        self.assertEqual(ssh.call_count, 1)

    def test_existing_credentials_are_not_requested(self):
        with patch.object(launcher, 'ssh', return_value=subprocess.CompletedProcess([], 0, '["TOKEN"]')) as ssh:
            with patch.object(launcher.getpass, 'getpass', side_effect=AssertionError('Unexpected prompt')):
                launcher.credentials(self.state(), [{'name': 'TOKEN', 'description': 'test'}])
        self.assertEqual(ssh.call_count, 1)


class EnvironmentWriterTests(unittest.TestCase):
    def test_literal_values_and_existing_credentials_survive_reapplication(self):
        spec = importlib.util.spec_from_file_location('environment_writer', Path(__file__).parent / 'files/environment.py')
        writer = importlib.util.module_from_spec(spec)
        spec.loader.exec_module(writer)
        with tempfile.TemporaryDirectory() as tmp:
            home = Path(tmp)
            directory = home / '.config/devwright'
            directory.mkdir(parents=True)
            credentials = directory / 'credentials.sh'
            credentials.write_text("export EXISTING='preserve'\n")
            writer.configure(home, {'LITERAL': "a 'quote', $HOME, and $(false)"})
            writer.configure(home, {'LITERAL': "a 'quote', $HOME, and $(false)"})
            self.assertEqual(credentials.read_text().count('environment.sh'), 1)
            self.assertIn("export EXISTING='preserve'", credentials.read_text())
            self.assertEqual((directory / 'environment.sh').stat().st_mode & 0o777, 0o600)
            result = subprocess.run(['bash', '-c', 'source "$1"; printf "%s" "$LITERAL"', 'bash', str(directory / 'environment.sh')],
                                    text=True, capture_output=True, check=True)
            self.assertEqual(result.stdout, "a 'quote', $HOME, and $(false)")


if __name__ == '__main__':
    unittest.main()

File: environment/test_launcher.rb
# frozen_string_literal: true
# Standard-library-only tests: ruby environment/test_launcher.rb
require_relative 'launcher'

def assert(value, message = 'Assertion failed')
  raise message unless value
end

def rejects
  begin
    yield
  rescue StandardError
    return
  end
  raise 'Expected rejection'
end

count = 0
test = lambda do |name, &block|
  block.call
  count += 1
  puts "PASS #{name}"
end

Dir.mktmpdir('lima-recipe-tests-') do |dir|
  path = File.join(dir, 'config.yaml')
  test.call('unknown and removed recipe options rejected') do
    [{ 'ansible_connection' => 'local' }, { 'replace_codex_config' => true }, { 'environment' => { 'BAD-NAME' => 'x' } }, { 'dotfiles_install' => '../escape' }, { 'packages' => ['--bad'] }, { 'credentials' => [{ 'name' => 'TOKEN', 'description' => 'test' }], 'environment' => { 'TOKEN' => 'secret' } }].each do |data|
      File.write(path, YAML.dump(data))
      rejects { Environment.config(path) }
    end
  end
  test.call('script paths resolved and Bash syntax checked') do
    File.write(File.join(dir, 'setup.sh'), "#!/bin/bash\necho ready\n")
    File.write(path, YAML.dump({ 'user_script' => 'setup.sh' }))
    assert(Environment.config(path)['user_script'] == File.join(dir, 'setup.sh'))
    File.write(File.join(dir, 'setup.sh'), "if then\n")
    rejects { Environment.config(path) }
  end
  data = Environment.config(File.join(__dir__, 'config.yaml'))
  vm = Environment.render(data)
  script = vm['provision'][0]['script']
  test.call('render contains one Bash provisioner with a completion guard') do
    assert(vm['provision'].length == 1)
    assert(script.index('lima-setup-complete') < script.index('base64 -d'))
    assert(Digest::SHA256.hexdigest(script) == JSON.parse(vm['message'])['environment_script_sha256'])
    Environment.capture('bash', '-n', input: script)
  end
  test.call('failed setup retries and completed setup is skipped') do
    test_state = File.join(dir, 'boot-state')
    payload = "#!/bin/bash\nset -eu\necho attempt >> #{File.join(dir, 'attempts').shellescape}\nif [ ! -e #{File.join(dir, 'retry').shellescape} ]; then touch #{File.join(dir, 'retry').shellescape}; exit 23; fi\n"
    guard = Environment.read('environment/provision.sh').sub('__SETUP_B64__', Environment.encoded(payload)).sub('state=/usr/local/share/devwright', "state=#{test_state.shellescape}")
    rejects { Environment.capture('bash', input: guard) }
    assert(!File.exist?(File.join(test_state, 'lima-setup-complete')))
    Environment.capture('bash', input: guard)
    Environment.capture('bash', input: guard)
    assert(File.readlines(File.join(dir, 'attempts')).length == 2)
    assert(File.exist?(File.join(test_state, 'lima-setup-complete')))
  end
  state = { 'name' => 'test', 'dir' => File.join(dir, 'test'), 'config' => vm }
  test.call('effective recipe accepted and injected provisions rejected') do
    Environment.check_instance(state, 'test', expected: vm)
    other = Marshal.load(Marshal.dump(state))
    other['config']['provision'] << { 'mode' => 'system', 'script' => 'bad' }
    rejects { Environment.check_instance(other, 'test') }
  end
  test.call('global overrides rejected before first boot') do
    %w[mounts portForwards probes].each do |key|
      other = Marshal.load(Marshal.dump(state))
      other['config'][key] = [{}]
      rejects { Environment.check_instance(other, 'test', expected: vm) }
    end
    other = Marshal.load(Marshal.dump(state))
    other['config']['provision'][0]['script'] += "\necho injected\n"
    rejects { Environment.check_instance(other, 'test') }
  end
  test.call('SSH aliases preserve existing host configuration and separate users') do
    FileUtils.mkdir_p(File.join(dir, '.ssh'))
    File.write(File.join(dir, '.ssh/config'), "Host personal\n  HostName example.com\n")
    2.times { Environment.install_ssh(state, home: dir) }
    text = File.read(File.join(dir, '.ssh/config'))
    assert(text.scan('Include').length == 1 && text.include?('Host personal'))
    assert(Environment.ssh_config(state).include?('ControlPath ~/.ssh/control-%C'))
    %w[root dev].each do |user|
      args = Environment.ssh_args(state, user, ['echo', '$(literal)'])
      assert(args.include?('IdentityAgent=none') && args[args.index('-l') + 1] == user)
      assert(args.last == ['echo', '$(literal)'].shelljoin)
    end
  end
  test.call('credential values stay in stdin script and are shell-quoted') do
    value = "literal '\" $HOME $(false)\nsecond line"
    text = Environment.credential_script({ 'TOKEN' => value })
    Environment.capture('bash', '-n', input: text)
    encoded = text.match(/printf '%s' '([^']+)'/)[1]
    assignment = Base64.strict_decode64(encoded)
    output = Environment.capture('bash', '-c', assignment + '\nprintf "%s" "$TOKEN"'.sub('\\n', "\n"))
    assert(output == value)
    rejects { Environment.credential_script({ 'BAD-NAME' => 'x' }) }
  end
  test.call('configuration-in-place explicitly rejected') do
    rejects { Environment.main(['configure', 'test']) }
  end
end
puts "#{count} tests passed"
